### ISSUE TITLE
Can't pickle local object 'get_collate_fn.<locals>.collate_fn' after 4c71886

### ISSUE THREAD

**Issue by Valentine93** [CCRcmcpe/scal-sdt#7]
Really great project! I have been using scal-sdt for weeks. Happy to see the updates and LoRA support!

OS: Windows 10
Python: 3.10.6, 3.10.9 (Same error)

After 4c71886b2a47fdc3299c6e177008290ab3394d19 (for example at f62fb5fb5a933891339144801467e777e40554fb), I can no longer conduct training (even using the default `dreambooth.yaml`). Here is a stacktrace:
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Python\Python310\lib\multiprocessing\spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "C:\Python\Python310\lib\multiprocessing\spawn.py", line 126, in _main
    self = reduction.pickle.load(from_parent)
EOFError: Ran out of input
Traceback (most recent call last):
  File "C:\scal-sdt\train.py", line 143, in <module>
    main(*get_params())
  File "C:\scal-sdt\train.py", line 138, in main
    trainer.fit(model=model, ckpt_path=args.resume)
  File "C:\scal-sdt\venv\lib\site-packages\pytorch_lightning\trainer\trainer.py", line 603, in fit
    call._call_and_handle_interrupt(
  File "C:\scal-sdt\venv\lib\site-packages\pytorch_lightning\trainer\call.py", line 38, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "C:\scal-sdt\venv\lib\site-packages\pytorch_lightning\trainer\trainer.py", line 645, in _fit_impl
    self._run(model, ckpt_path=self.ckpt_path)
  File "C:\scal-sdt\venv\lib\site-packages\pytorch_lightning\trainer\trainer.py", line 1098, in _run
    results = self._run_stage()
  File "C:\scal-sdt\venv\lib\site-packages\pytorch_lightning\trainer\trainer.py", line 1177, in _run_stage
    self._run_train()
  File "C:\scal-sdt\venv\lib\site-packages\pytorch_lightning\trainer\trainer.py", line 1200, in _run_train
    self.fit_loop.run()
  File "C:\scal-sdt\venv\lib\site-packages\pytorch_lightning\loops\loop.py", line 199, in run
    self.advance(*args, **kwargs)
  File "C:\scal-sdt\venv\lib\site-packages\pytorch_lightning\loops\fit_loop.py", line 267, in advance
    self._outputs = self.epoch_loop.run(self._data_fetcher)
  File "C:\scal-sdt\venv\lib\site-packages\pytorch_lightning\loops\loop.py", line 194, in run
    self.on_run_start(*args, **kwargs)
  File "C:\scal-sdt\venv\lib\site-packages\pytorch_lightning\loops\epoch\training_epoch_loop.py", line 161, in on_run_start
    _ = iter(data_fetcher)  # creates the iterator inside the fetcher
  File "C:\scal-sdt\venv\lib\site-packages\pytorch_lightning\utilities\fetching.py", line 179, in __iter__
    self._apply_patch()
  File "C:\scal-sdt\venv\lib\site-packages\pytorch_lightning\utilities\fetching.py", line 120, in _apply_patch
    apply_to_collections(self.loaders, self.loader_iters, (Iterator, DataLoader), _apply_patch_fn)
  File "C:\scal-sdt\venv\lib\site-packages\pytorch_lightning\utilities\fetching.py", line 156, in loader_iters
    return self.dataloader_iter.loader_iters
  File "C:\scal-sdt\venv\lib\site-packages\pytorch_lightning\trainer\supporters.py", line 555, in loader_iters
    self._loader_iters = self.create_loader_iters(self.loaders)
  File "C:\scal-sdt\venv\lib\site-packages\pytorch_lightning\trainer\supporters.py", line 595, in create_loader_iters
    return apply_to_collection(loaders, Iterable, iter, wrong_dtype=(Sequence, Mapping))
  File "C:\scal-sdt\venv\lib\site-packages\lightning_utilities\core\apply_func.py", line 47, in apply_to_collection
    return function(data, *args, **kwargs)
  File "C:\scal-sdt\venv\lib\site-packages\torch\utils\data\dataloader.py", line 439, in __iter__
    self._iterator = self._get_iterator()
  File "C:\scal-sdt\venv\lib\site-packages\torch\utils\data\dataloader.py", line 390, in _get_iterator
    return _MultiProcessingDataLoaderIter(self)
  File "C:\scal-sdt\venv\lib\site-packages\torch\utils\data\dataloader.py", line 1077, in __init__
    w.start()
  File "C:\Python\Python310\lib\multiprocessing\process.py", line 121, in start
    self._popen = self._Popen(self)
  File "C:\Python\Python310\lib\multiprocessing\context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "C:\Python\Python310\lib\multiprocessing\context.py", line 336, in _Popen
    return Popen(process_obj)
  File "C:\Python\Python310\lib\multiprocessing\popen_spawn_win32.py", line 93, in __init__
    reduction.dump(process_obj, to_child)
  File "C:\Python\Python310\lib\multiprocessing\reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
AttributeError: Can't pickle local object 'get_collate_fn.<locals>.collate_fn'. Did you mean: '_loader_iters'?
```

I see you changed `modules\dataset\__init__.py:57 collate_fn()` in f62fb5fb5a933891339144801467e777e40554fb to `get_collate_fn()` which dynamically produces `collate_fn()` that you passed to `modules\model.py:292 torch.utils.data.DataLoader()` (https://github.com/CCRcmcpe/scal-sdt/commit/f62fb5fb5a933891339144801467e777e40554fb#diff-6b241d614aff2f2c6110a0c2b9482fcd3e223372030ad32603803d8db3f471b8R57). I found https://stackoverflow.com/a/52283968 which suggested that `multiprocessing` does not support pickling nested functions. Indeed, unnesting `get_collate_fn()` to `collate_fn_cache()` and `collate_fn_no_cache()` and passing the unnested functions to `torch.utils.data.DataLoader()` can fix this issue.

**Comment by CCRcmcpe** [comment]
Should be fixed with [3e603de](https://github.com/CCRcmcpe/scal-sdt/commit/3e603de701a9ae08a21663a4953faac620faf51a). I don't have a windows machine to test it, feel free to comment if not.

**Comment by Valentine93** [comment]
It works perfectly! Thank you so much!
